pyro/contrib/gp/util.py

```diff
@@ -128,3 +130,82 @@ def _register_param(self, param, mode="model"):
             p = pyro.sample(param_name, dist.Delta(MAP_param))
 
         self._registered_params[param] = p
+
+
+class LowerConfidenceBound(nn.Module):
+    """
+    A simple implementation of Lower Confidence Bound acquisition function.
+    This class is useful for doing Bayesian Optimization.
+
+    :param pyro.contrib.gp.Model model: A Gaussian Process model which is used
+        to approximate an objective function.
+    :param torch.distributions.constraints.interval constraint: A constraint which forces
+        input variable stay in an interval.
+    :param int num_restarts: Number of times we want to find minimum points of current acquisition
+        posterior. The higher it is, we get a better minimum point.
+    :param int num_steps: We use Adam optimizer with ``lr=1`` to find a mimimum point of current
+        acquisition posterior. The number of iterations is ``num_steps``.
+    """
+    def __init__(self, model, constraint, num_restarts=5, num_steps=1000):
+        super(LowerConfidenceBound, self).__init__()
+        self.model = model
+        self.constraint = constraint
+        self.num_restarts = num_restarts
+        self.num_steps = num_steps
+        self._x0 = None
+
+    def _get_init_point(self, random=True):
+        if not random and self._x0 is not None:
+            return self._x0
+        else:
+            return self.model.X.new(1).uniform_(self.constraint.lower_bound,
+                                                self.constraint.upper_bound)
+
+    def update_data(self, x, objective):
+        """
+        Updates Gaussian Process data and optimizes it.
+        """
+        self._x0 = x
+        y = objective(x)
+        X = torch.cat([self.model.X, x])
+        y = torch.cat([self.model.y, y])
+        self.model.set_data(X, y)
+        self.model.optimize(optimizer=Adam({}))
+
+    def minimum(self):
+        """
+        Gets minimum points of current acquisition posterior.
+        """
+        candidates = []
+        values = []
+        for i in range(self.num_restarts):
+            # last minimum point will be an init point for first minimum candidate,
+            # other minimum candidates will get uniform random initialization
+            unconstrained_x0 = transform_to(self.constraint).inv(self._get_init_point(random=(i != 0)))
```

Comment:
This solution will not scale well in higher dimensional BO, maybe allow users to pass in different initialization strategies?

Reply:
@ysaatchi This class was created for just this particular tutorial. For a BO module with various features, @tobyclh is working on it.

Reply:
got it, cool :)